rust/hyperlane-base/src/contract_sync/cursor.rs

```diff
@@ -36,6 +36,11 @@ pub(crate) struct MessageSyncCursor {
     /// The next block that should be indexed.
     next_block: u32,
     /// The next nonce that the cursor is looking for.
+    /// In the EVM, this is used for optimizing indexing,
+    /// because it's cheaper to make read calls for the nonce than
+    /// to call `eth_getLogs` with a block range.
+    /// In Sealevel, historic queries aren't supported, so the nonce field
+    /// is used to query storage in sequence.
     next_nonce: u32,
```

Comment:
So i think you may have had this idea and either I'm forgetting what we didn't like about it, I didn't understand it correctly at the time, or I'm imagining things - so atm we have the MessageSyncCursor which already has all the logic to power either block-based indexing or sequence-based indexing, and it just so happens that it's built specifically for getting message nonces / its interface into the db is message specific (i.e. HyperlaneMessageStore)

Is there maybe an opportunity to maybe make this style of cursor less message-specific, so that we can use this type of cursor for IGP payments or message delivery indexing when the chain we're indexing supports the sequence mode? And then for IGP payment or message delivery indexing for chains like Ethereum that *don't* support sequence mode, we can still use the `RateLimitedContractSyncCursor`, but it only works with block ranges

If this is too out there / doesn't make sense I'm also happy with the path you're going down, just wanted to share

Reply:
I was considering this (e.g. using a supertrait to share logic between `MessageSyncCursor` and `RateLimitedContractSyncCursor`) as a follow up step once I get an initial working version, but could also try making it work at this stage. 


> for chains like Ethereum that don't support sequence mode, we can still use the RateLimitedContractSyncCursor

Isn't the point of #2511 to add sequence indexing to `RateLimitedContractSyncCursor` for IGP and message delivery, even if that means adding empty stubs to the EVM trait implementations? Is your suggestion to reuse the `MessageSyncCursor` on chains where sequence mode can be used?